### PR TITLE
boot.rst: add First Boot Determination section

### PR DESCRIPTION
**Pull request by OddBloke**
LP: #1888858

### CODE REVIEW

File: doc/rtd/topics/boot.rst [hunk 1]
```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
 scripts until cloud-init is done without having to write your own systemd
 units dependency chains. See :ref:`cli_status` for more info.
 
+First Boot Determination
+************************
+
+cloud-init has to determine whether or not the current boot is a first boot or
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:163]: "is a first boot" -> "is the first boot of a new instance".

File: doc/rtd/topics/boot.rst [hunk 2]
```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+stages and boots.
+
+If this cache is present, then cloud-init has run on this system before.
+[#not-present]_  There are two cases where this could occur.  Most commonly,
+the instance has been rebooted, and this is a second/subsequent boot.
+Alternatively, the filesystem has been attached to a *new* instance, and this
+is a first boot.  The most obvious case where this happens is when an instance
+is launched from an image captured from a launched instance.
+
+By default, cloud-init attempts to determine which case we're running in by
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:179]: we're -> "it is"

File: doc/rtd/topics/boot.rst [hunk 3]
```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+checking the instance ID in the cache against the instance ID it determines at
+runtime.  If they do not match, then this is a first boot; otherwise, it's a
+subsequent boot.  Internally, cloud-init refers to this behavior as ``check``.
+
+This behavior is required for images captured from launched instances to
+behave correctly, and so is the default which generic cloud images ship with.
+However, there are cases where it can cause problems. [#problems]_ For these
+cases, cloud-init has support for modifying its behavior to trust the instance
+ID that is present in the system unconditionally.  This means that cloud-init
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:188]: system -> cache

File: doc/rtd/topics/boot.rst [hunk 4]
```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+will never detect a first boot when the cache is present, and it follows that
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:189]: "first boot" -> "new instance" ?

File: doc/rtd/topics/boot.rst [hunk 5]
```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+the only way to cause cloud-init to detect a first boot is to manually remove
+cloud-init's cache.  Internally, this behavior is referred to as ``trust``.
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:191]: I'd just scrap "Internally....".
This is user-facing documentation.

File: doc/rtd/topics/boot.rst [hunk 6]
```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+cloud-init looks in ``/var/lib/cloud/instance/obj.pkl`` for its cache.
+``/var/lib/cloud/instance`` is a symlink to a subdirectory of
+``/var/lib/cloud/instances/``, which contains a single directory per instance
+that has booted with this filesystem.  To clean the cache in the least
+destructive manner possible, simply remove the ``/var/lib/cloud/instance``
+symlink.
+
+.. note:: If you do want to destroy all cached data, then in addition to
+   removing the ``/var/lib/cloud/instance`` symlink, you should also remove
+   all directories under ``/var/lib/cloud/instances``.
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:211]: I don't really like documenting this "interface".

for "all cached data", this should just be `cloud-init clean`.

I'm not sure if we need to mention to a user the ability to just remove the symlink or not.  Can you think of any non-developer use where that is valuable?

**Review by smoser — COMMENTED** [review]
First, thanks for writing this. Second, 2 general comments, with some in-line:

 1. "first boot", I think really should be consistently "first instance boot" or "first boot of a new instance".  cloud-init *does* identify a "first boot" (per-once) in addition to "first instance boot".
 2. I'd recommend dropping documentation of internal behavior that we do not necessarily wish to support.  Instead, provide a cloud-init sub-command that will do what you want.

2 things you could do:
 * cloud-init status could show if manual_cache_clean was set 
 * some sub-command could toggle the status of manual_cache_clean.

File: doc/rtd/topics/boot.rst (hunk 6 above)
Comment (OddBloke) [reply doc/rtd/topics/boot.rst:211]: Agreed, I've just changed it to reference clean.

File: doc/rtd/topics/boot.rst (hunk 5 above)
Comment (OddBloke) [reply doc/rtd/topics/boot.rst:191]: Scrap the whole sentence, or just "Internally, "?

(If the former: I would like to keep the names of the behaviours, because I think it makes it easier to know where to refer back to when you're reading the later parts of the documentation, and "in `trust` mode" is more concise than "has `manual_cache_clean` set to True".)

**Comment by OddBloke** [comment]
Thanks for the review Scott, this is much improved now. I believe I've addressed all your comments except for the one I've left unresolved.

**Review by lucasmoura — COMMENTED** [review]
I have no problem with the provided configuration, but just a question regarding instance-id. For EC2, isn't it possible for the `instance-id` to change without requiring cloud-init to reset the setting of the machine ? For example, if we change the region of an instance I think the `instance-id` will change, but would we like for cloud-init to reset the setting in that scenario ?

**Comment by smoser** [comment]
I don't think it's possible to move an instance in aws.

https://aws.amazon.com/premiumsupport/knowledge-center/move-ec2-instance/

File: doc/rtd/topics/boot.rst [hunk 7]
```diff
@@ -157,4 +157,89 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+However, there are cases where it can cause problems. [#problems]_ For these
+cases, cloud-init has support for modifying its behavior to trust the instance
+ID that is present in the system unconditionally.  This means that cloud-init
+will never detect a new instance when the cache is present, and it follows that
+the only way to cause cloud-init to detect a new instance (and therefore its
+first boot) is to manually remove cloud-init's cache.  Internally, this
+behavior is referred to as ``trust``.
+
+To configure which of these behaviors to use, cloud-init exposes the
+``manual_cache_clean`` configuration option.  When ``false``, cloud-init will
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:197]: When ``false`` (the default)

^ mention here that that is the default?

File: doc/rtd/topics/boot.rst [hunk 8]
```diff
@@ -157,4 +157,89 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+
+   This is a functional issue, but also a potential security one: cloud-init is
+   responsible for rotating SSH host keys on first boot, and this will not
+   happen on these instances.
+
+.. [#not-present] It follows that if this cache is not present, cloud-init has
+   not run on this system before, so this is unambiguously this instance's
+   first boot.
+
+.. [#problems] A couple of ways in which this has been observed to cause
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:229]: suggestion: A couple of ways in which strict reliance on presence of a datasource has been observed to cause problems:

File: doc/rtd/topics/boot.rst [hunk 9]
```diff
@@ -157,4 +157,89 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+
+    * if a cloud's metadata service is flaky, cloud-init's instance ID
+      determination will sometimes fail to determine the current instance ID,
+      which makes it impossible to determine if this is an instance's first or
+      subsequent boot (`#1885527`_)
+    * if cloud-init is used to provision a physical appliance or device using
+      the NoCloud datasource: if an attacker can present a NoCloud drive to the
+      device with a different instance ID, then cloud-init's default behavior
+      will detect this as an instance's first boot and reset the device using
+      the attacker's configuration (`#1879530`_)
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:240]: This isn't limited to instances currently using NoCloud datasource.  Nor is the attack vector limited to a NoCloud drive.

**Review by smoser — CHANGES_REQUESTED** [review]
some minor cleanups requested. this does look much better.

File: doc/rtd/topics/boot.rst [hunk 10]
```diff
@@ -157,4 +157,91 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+an instance is launched from an image captured from a launched instance.
+
+By default, cloud-init attempts to determine which case it is running in by
+checking the instance ID in the cache against the instance ID it determines at
+runtime.  If they do not match, then this is an instance's first boot;
+otherwise, it's a subsequent boot.  Internally, cloud-init refers to this
+behavior as ``check``.
+
+This behavior is required for images captured from launched instances to
+behave correctly, and so is the default which generic cloud images ship with.
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:187]: Maybe reword like:

     This behavior is the default which generic images ship with.  It is required in order to avoid a manual clean operation before image capture.

then you introduce the word 'manual', and possibly the name 'manual_cache_clean' makes some more sense.

Comment (smoser) [reply doc/rtd/topics/boot.rst:187]: but i see you did similarly below... take the above only as a suggestion.

File: doc/rtd/topics/boot.rst [hunk 11]
```diff
@@ -157,4 +157,91 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+   happen on these instances.
+
+.. [#not-present] It follows that if this cache is not present, cloud-init has
+   not run on this system before, so this is unambiguously this instance's
+   first boot.
+
+.. [#problems] A couple of ways in which this strict reliance on the presence
+   of a datasource has been observed to cause problems:
+
+    * If a cloud's metadata service is flaky, cloud-init's instance ID
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:232]: "is flaky and cloud-init cannot obtain the instance-id locally on that platform, cloud-init's instance ID...."

just to point out that we do try to do sane things.

File: doc/rtd/topics/boot.rst [hunk 12]
```diff
@@ -157,4 +157,91 @@ finished, the ``cloud-init status`` subcommand can help block external
...
+      determination will sometimes fail to determine the current instance ID,
+      which makes it impossible to determine if this is an instance's first or
+      subsequent boot (`#1885527`_).
+    * If cloud-init is used to provision a physical appliance or device using
+      the NoCloud datasource: if an attacker can present a NoCloud drive to the
```
Comment (smoser) [inline doc/rtd/topics/boot.rst:237]: i think just "present a datasource".  If you want, you can mention "such as NoCloud or ConfigDrive"...

**Review by smoser — APPROVED** [review]
